File: 백준/Gold/1256. 사전/사전.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    private static final int MAX = 1_000_000_000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        int[][] combination = new int[n + m + 1][n + m + 1];
        for (int i = 0; i <= n + m; i++) {
            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    combination[i][j] = 1;
                } else {
                    combination[i][j] = combination[i - 1][j - 1] + combination[i - 1][j];

                    if (combination[i][j] > MAX) {
                        combination[i][j] = MAX + 1;
                    }
                }
            }
        }

        if (combination[n + m][m] < k) {
            System.out.println(-1);
            return;
        }

        StringBuilder sb = new StringBuilder();
        while (n > 0 || m > 0) {
            if (n == 0) {
                sb.append('z');
                m--;
            } else if (m == 0) {
                sb.append('a');
                n--;
            } else {
                int count = combination[n - 1 + m][m];

                if (count >= k) {
                    sb.append('a');
                    n--;
                } else {
                    sb.append('z');
                    k -= count;
                    m--;
                }
            }
        }

        System.out.println(sb);
    }
}

File: 백준/Gold/1256. 사전/사전.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private const val max = 1_000_000_000

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m, k) = br.readLine().split(" ").map { it.toInt() }
    val combination = Array(n + m + 1) { IntArray(n + m + 1) }

    for (i in 0..n + m) {
        for (j in 0..i) {
            combination[i][j] = when {
                j == 0 || j == i -> 1
                else -> {
                    val sum = combination[i - 1][j - 1] + combination[i - 1][j]
                    if (sum > max) max + 1 else sum
                }
            }
        }
    }

    if (combination[n + m][m] < k) {
        println(-1)
        return
    }

    var remainA = n
    var remainZ = m
    var remainK = k

    val sb = StringBuilder()
    while (remainA > 0 || remainZ > 0) {
        when {
            remainA == 0 -> {
                sb.append('z')
                remainZ--
            }
            remainZ == 0 -> {
                sb.append('a')
                remainA--
            }
            else -> {
                val count = combination[remainA - 1 + remainZ][remainZ]

                if (count >= remainK) {
                    sb.append('a')
                    remainA--
                } else {
                    sb.append('z')
                    remainK -= count
                    remainZ--
                }
            }
        }
    }

    println(sb)
}
